src/engine/sync/abletonlink.cpp

```diff
@@ -0,0 +1,217 @@
+    m_syncMode = syncMode;
+}
+
+void AbletonLink::notifyUniquePlaying() {
+}
+
+void AbletonLink::requestSync() {
+    qDebug() << "AbletonLink::requestSync()";
+}
+
+SyncMode AbletonLink::getSyncMode() const {
+    return m_syncMode;
+}
+
+bool AbletonLink::isPlaying() const {
+    if (!m_pLink->isEnabled()) {
+        return false;
+    }
+    if (m_pLink->numPeers() < 1) {
+        return false;
+    }
+
+    // Note, that ableton::Link::SessionState.isPlaying() is an optional Ableton
+    // Link feature (Start/Stop sync) and shouldn't be taken into account here.
+    return true;
+}
+bool AbletonLink::isAudible() const {
+    return false;
+}
+bool AbletonLink::isQuantized() const {
+    return true;
+}
+
+mixxx::Bpm AbletonLink::getBpm() const {
+    return getBaseBpm();
+}
+
+double AbletonLink::getBeatDistance() const {
+    const auto sessionState = m_pLink->captureAudioSessionState();
+    const auto beats = sessionState.beatAtTime(
+            m_absTimeWhenPrevOutputBufferReachesDac, getQuantum());
+    return std::fmod(beats, 1.0);
+}
+
+mixxx::Bpm AbletonLink::getBaseBpm() const {
+    const auto sessionState = m_pLink->captureAudioSessionState();
+    return mixxx::Bpm(sessionState.tempo());
+}
+
+void AbletonLink::updateLeaderBeatDistance(double beatDistance) {
+    auto sessionState = m_pLink->captureAudioSessionState();
+    const auto currentBeat = sessionState.beatAtTime(
+            m_absTimeWhenPrevOutputBufferReachesDac, getQuantum());
+    const auto newBeat = currentBeat - std::fmod(currentBeat, 1.0) + beatDistance;
+
+    sessionState.requestBeatAtTime(newBeat, m_absTimeWhenPrevOutputBufferReachesDac, getQuantum());
+    m_pLink->commitAudioSessionState(sessionState);
+}
+
+void AbletonLink::forceUpdateLeaderBeatDistance(double beatDistance) {
+    auto sessionState = m_pLink->captureAudioSessionState();
+    const auto currentBeat = sessionState.beatAtTime(
+            m_absTimeWhenPrevOutputBufferReachesDac, getQuantum());
+    const auto newBeat = currentBeat - std::fmod(currentBeat, 1.0) + beatDistance;
+
+    sessionState.forceBeatAtTime(newBeat, m_absTimeWhenPrevOutputBufferReachesDac, getQuantum());
+    m_pLink->commitAudioSessionState(sessionState);
+}
+
+void AbletonLink::updateLeaderBpm(mixxx::Bpm bpm) {
+    auto sessionState = m_pLink->captureAudioSessionState();
+    sessionState.setTempo(bpm.value(), m_absTimeWhenPrevOutputBufferReachesDac);
+    m_pLink->commitAudioSessionState(sessionState);
+}
+
+void AbletonLink::notifyLeaderParamSource() {
+    // In Ableton Link all pears are equal. Therefore nothing differs,
+    // if AbletonLink becomes SyncLeader.
+    // TODO: Check the special case of half/double BPM sync.
+}
+
+void AbletonLink::reinitLeaderParams(double beatDistance, mixxx::Bpm baseBpm, mixxx::Bpm bpm) {
+    Q_UNUSED(baseBpm)
+    updateLeaderBeatDistance(beatDistance);
+    updateLeaderBpm(bpm);
+}
+
+void AbletonLink::updateInstantaneousBpm(mixxx::Bpm bpm) {
+    Q_UNUSED(bpm)
+}
```

Comment:
[CppCoreGL F.9](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f9-unused-parameters-should-be-unnamed)
```suggestion
void AbletonLink::reinitLeaderParams(double beatDistance, mixxx::Bpm, mixxx::Bpm bpm) {
    updateLeaderBeatDistance(beatDistance);
    updateLeaderBpm(bpm);
}

void AbletonLink::updateInstantaneousBpm(mixxx::Bpm) {
}
```